example/neumann-reduced/heat_equation_reduced.py

```diff
@@ -61,42 +28,55 @@ class StationaryPreciceModel(Model):
     ----------
     operator
         The |Operator| of the linear problem.
-    coupling_operator
-        Operator handling the coupling contributions and the exchange with preCICE
+    coupling_input_operator
+        Operator mapping the coupling contributions to the rhs.
+    coupling_output_operator
+        Operator mapping the solution to the coupling contributions.
     """
 
-    def __init__(self, operator, coupling_operator,
-                 output_functional=None, products=None,
-                 error_estimator=None, visualizer=None, name=None):
+    def __init__(self, operator,
+                 coupling_input_operator=None, coupling_output_operator=None,
+                 output_functional=None, products=None, error_estimator=None,
+                 visualizer=None, name=None):
 
-        assert output_functional is None or output_functional.source == operator.source
+        coupling_input_operator = coupling_input_operator or IdentityOperator(operator.range)
+        coupling_output_operator = coupling_output_operator or IdentityOperator(operator.source)
+        output_functional = output_functional or ZeroOperator(NumpyVectorSpace(0), operator.source)
+        assert coupling_input_operator.range == operator.range
+        assert coupling_output_operator.source == operator.source
+        assert output_functional.source == operator.source
 
         super().__init__(products=products, error_estimator=error_estimator,
                          visualizer=visualizer, name=name)
 
         self.__auto_init(locals())
         self.solution_space = operator.source
-        if output_functional is not None:
-            self.dim_output = output_functional.range.dim
-
-    _compute_allowed_kwargs = frozenset({'coupling_input', 'coupling_output'})
+        self.dim_output = output_functional.range.dim
 
-    def _compute_solution(self, mu=None, **kwargs):
-        # this should only be called when compute has been called with
-        # solution=False and e.g. output=True
-        return self._compute(solution=True, mu=mu, **kwargs)['solution']
-
-    def _compute(self, solution, mu=None, **kwargs):
-        retval = {}
+    _compute_allowed_kwargs = frozenset({'coupling_input'})
 
+    def _compute_solution(self, mu=None, coupling_input=None, **kwargs):
         # Assemble the RHS originating from the coupling data
-        rhs = self.coupling_operator.apply_assemble(solution)
+        rhs = self.coupling_input_operator.apply(coupling_input)
```

Comment:
@sdrave I will have a closer look at the code (hopefully) next week. To recap our discussion: overall goal is to make the operator stationary so that we can get rid of the RHS time-dependency. What was again the desired change in the python code? You wanted to get rid of this line and fuse it into the precice advance, right?

Reply:
Sorry for the late reply! 

- I suggested to solve a stationary PDE instead of a time-dependent problem as I think this would make things conceptually simpler. Usually, for a time-dependent problem the time stepping would be done inside the `Model`'s solve method. That would mean that the `Model` would need to manage what is now called `PreciceCoupler`. For a stationary problem, the `Model` does not need to know anything about how to communicate with the outside world. Also, an elliptic problem is somewhat easier to deal with mathematically.

- Regarding `coupling_input_operator`: This is currently an `IdentityOperator` and does nothing at all. The idea would be that it should do the rhs assembly for given data as communicated by precice. For a stationary problem, one should be able to refactor `assemble_rhs` in a way that the resulting operator becomes linear. (There will be a fixed rhs not depending on the coupling data that needs to be added.) Then, this code can be reduced as well s.t. no deal.ii code needs to run in the reduced model.

- It would be beneficial to have a way to perform multiple coupled simulations (for different parameters) in one program run.

Reply:
@DavidSCN I'm not sure how clear I expressed myself. As I said, we can also setup another call to discuss this in more detail.

Reply:
> For a stationary problem, the Model does not need to know anything about how to communicate with the outside world.

In order to solve the stationary problem on one side -- no, the Model does not need to know anything about how to communicate data. In a coupled sense, we would still have multiple iterations between all solvers in order to converge with the coupled system towards a solution, right? Then, we would still have the `PreciceCoupler` (doing less work in the assembly).

> For a stationary problem, one should be able to refactor assemble_rhs in a way that the resulting operator becomes linear.

We receive a flux at the coupling interface and I would reduce the current RHS assembly into an integration of the flux at the coupling boundary only. This is the most simple version I can think of.

> It would be beneficial to have a way to perform multiple coupled simulations (for different parameters) in one program run.

This requires probably a bit more discussion.

Reply:
> > For a stationary problem, the Model does not need to know anything about how to communicate with the outside world.
> 
> In order to solve the stationary problem on one side -- no, the Model does not need to know anything about how to communicate data. 

I would actually start with stationary on both sides if that is no major problem. 

> In a coupled sense, we would still have multiple iterations between all solvers in order to converge with the coupled system towards a solution, right? 

Yes.

>Then, we would still have the `PreciceCoupler` (doing less work in the assembly).

Yes, but leaving it outside the model would be more natural.

> 
> > For a stationary problem, one should be able to refactor assemble_rhs in a way that the resulting operator becomes linear.
> 
> We receive a flux at the coupling interface and I would reduce the current RHS assembly into an integration of the flux at the coupling boundary only. This is the most simple version I can think of.

The assembly code can be arbitrary complex. As long as it is linear in the communicated data and does not depend on previous state. So one would need to split whatever source term / boundary term there is into a separate method from the computation of the interface contribution.

> > It would be beneficial to have a way to perform multiple coupled simulations (for different parameters) in one program run.
> 
> This requires probably a bit more discussion.

If I remember correctly, Benjamin wanted to look into this? I would guess that it is also possible to fake this by just continuing to iterate and sending new data.